File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int count=0;

    public int goodNodes(TreeNode root) {
        checkGoodNodes(root,root.val);
        return count;
    }

    void checkGoodNodes(TreeNode root,int max){
         if(root == null)
            return;

        if(root.val >= max)
            count++;

            checkGoodNodes(root.left,Math.max(root.val,max));
            checkGoodNodes(root.right,Math.max(root.val,max));
    }
}

File: Data Structures & Algorithms/house-robber-iii/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        auto [rob,skip]=rob_house(root);
        return max(rob,skip);        
    }

    pair<int,int> rob_house(TreeNode* node){
         if(node==nullptr){
         return {0,0};   
        }
        int rob,skip;
        auto [left_rob,left_skip]=rob_house(node->left);
        auto [right_rob,right_skip]=rob_house(node->right);
        rob=node->val + left_skip +right_skip;
        skip=max(left_skip,left_rob) + max(right_skip,right_rob);
        return {rob,skip};

    }
};